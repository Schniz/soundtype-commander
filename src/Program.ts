import { Command } from 'commander';

export class Program<ArgumentTypes extends { [key: string]: any }> {
  private readonly parseFn: (cmd: Command) => ArgumentTypes;
  private readonly command: Command;

  constructor(command: Command, parse: (cmd: Command) => ArgumentTypes) {
    this.parseFn = parse;
    this.command = command;
  }

  /** Execute commander */
  parse(argv: string[]): ArgumentTypes {
    return this.parseFn(this.command.parse(argv));
  }
}
