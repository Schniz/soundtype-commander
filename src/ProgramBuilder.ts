import { Command } from 'commander';
import { Program } from './Program';

type If<
  Condition extends { Cond: boolean; True: any; False: any }
> = Condition['Cond'] extends true ? Condition['True'] : Condition['False'];

type VariadicArgsOptions<Name extends string = string> = {
  name: Name;
  required: boolean;
};

type OptionsFor<X> = {
  [key in keyof X]: {
    parse(str: string): X[key];
    default: X[key];
  };
};

export class ProgramBuilder<
  ArgumentTypes extends { [key: string]: any } = {},
  HasVariadic extends boolean = false
> {
  private readonly options: OptionsFor<ArgumentTypes>;
  private readonly variadicOpts: If<{
    Cond: HasVariadic;
    True: VariadicArgsOptions;
    False: undefined;
  }>;
  private readonly enhancers: ((cmd: Command) => Command)[] = [];
  private readonly name: string;
  private readonly version: string;

  constructor(
    enhancers: ((cmd: Command) => Command)[],
    options: OptionsFor<ArgumentTypes>,
    name: string,
    version: string,
    variadicOpts: If<{
      Cond: HasVariadic;
      True: VariadicArgsOptions;
      False: undefined;
    }>
  ) {
    this.enhancers = enhancers;
    this.options = options;
    this.name = name;
    this.version = version;
    this.variadicOpts = variadicOpts;
  }

  static create(name: string, version: string): ProgramBuilder<{}, false> {
    return new ProgramBuilder<{}>([], {}, name, version, undefined);
  }

  option<ArgumentName extends string, ArgumentType>(
    opts: {
      name: ArgumentName;
      shorthand?: string;
      default: ArgumentType;
      description: string;
    } & (ArgumentType extends boolean
      ? { parse?: never; argName?: never }
      : {
          parse(val: string): ArgumentType;
          argName?: string;
        })
  ) {
    const optionArgument =
      typeof opts.default === 'boolean' ? '' : ` <${opts.argName || 'arg'}>`;
    const shorthandDef = opts.shorthand ? `-${opts.shorthand}, ` : '';

    const parse = typeof opts.default === 'boolean' ? Boolean : opts.parse;

    return new ProgramBuilder<
      ArgumentTypes & { [key in ArgumentName]: ArgumentType },
      HasVariadic
    >(
      [
        ...this.enhancers,
        cmd =>
          cmd.option(
            `${shorthandDef}--${opts.name}${optionArgument}`,
            opts.description,
            opts.default
          ),
      ],
      {
        ...this.options,
        [opts.name]: { default: opts.default, parse },
      },
      this.name,
      this.version,
      this.variadicOpts
    );
  }

  variadic<Name extends string>(
    opts: If<{
      Cond: HasVariadic;
      True: never;
      False: VariadicArgsOptions<Name>;
    }>
  ): ProgramBuilder<ArgumentTypes & { [key in Name]: string[] }, true> {
    return new ProgramBuilder<
      ArgumentTypes & { [key in Name]: string[] },
      true
    >(this.enhancers, this.options, this.name, this.version, opts);
  }

  private createCommand(): Command {
    const command = new Command();
    const usageVariadic = !this.variadicOpts
      ? ''
      : this.variadicOpts.required
      ? ` <${this.variadicOpts.name}>`
      : ` [${this.variadicOpts.name}]`;
    command
      .name(this.name)
      .version(this.version)
      .usage(`[options]${usageVariadic}`);
    return this.enhancers.reduce((cmd, fn) => fn(cmd), command);
  }

  build(): Program<ArgumentTypes> {
    return new Program(this.createCommand(), p => {
      const result: ArgumentTypes = {} as any;

      if (this.variadicOpts) {
        result[this.variadicOpts.name as keyof ArgumentTypes] = p.args as any;
      }

      for (const [key, option] of Object.entries(this.options)) {
        result[key as keyof ArgumentTypes] =
          p[key] === undefined ? option.default : option.parse(p[key]);
      }

      return result;
    });
  }
}
