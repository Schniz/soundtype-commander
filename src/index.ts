import { ProgramBuilder } from './ProgramBuilder';

export const program = ProgramBuilder.create;
