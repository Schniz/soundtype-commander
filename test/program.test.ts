import { program } from '../src/index';
import { getTypes } from 'infer-types';
import * as path from 'path';

test('works', () => {
  const myProgram = program('my-app', '1.0.0')
    .option({
      name: 'bool',
      description: 'this is a boolean',
      default: false,
    })
    .option({
      name: 'num',
      description: 'this is a number',
      default: undefined,
      parse: str => {
        const value = parseInt(str, 10);
        if (Number.isNaN(value)) throw new Error('This is not an int');
        return value;
      },
    })
    .variadic({ name: 'positional', required: true })
    .build();

  const result = myProgram.parse([
    'node',
    'my-app.js',
    '--num=10',
    'positional',
    '--bool',
    'argument',
  ]);

  expect(result).toEqual({
    bool: true,
    num: 10,
    positional: ['positional', 'argument'],
  });
});

test('inference', () => {
  const inferredTypes = getTypes(path.join(__dirname, 'simple-program.ts'));
  expect(inferredTypes).toMatchSnapshot();
});
