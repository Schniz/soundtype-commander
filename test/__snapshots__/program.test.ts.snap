// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inference 1`] = `
Object {
  "result": "{ bool: boolean; } & { num: number; } & { positional: string[]; }",
}
`;
